T5518: Add basic MLD support

vfreex opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->

Currently VyOS has `protocol igmp` option to enable IGMP querier and reports through FRR's pimd.

I would like to add support for IPv6 as well since FRR's IPv6 multicast functionality has significantly improved.

Enabling both MLD and IGMP on a VyOS router will allow us to turn on multicast snooping on layer-3 switches in dual-stack networks.

Example commands:

```
// Enable on interface eth0
set protocols pim6 interface eth0

// Explicitly join multicast group ff18::1234 on interface eth1
set protocols pim6 interface eth1 mld join ff18::1234

// Explicitly join source-specific multicast group ff38::5678 with source address 2001:db8::1 on interface eth1
set protocols pim6 interface eth1 mld join ff38::5678 source 2001:db8::1
```
## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://vyos.dev/T5518

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->

## Proposed changes
<!--- Describe your changes in detail -->

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

vfreex committed:
T5518: Add basic MLD support

Currently VyOS has `protocol igmp` option to enable IGMP querier and reports through FRR's pimd.

I would like to add support for IPv6 as well since FRR's IPv6 multicast functionality has significantly improved.

Enabling both MLD and IGMP on a VyOS router will allow us to turn on multicast snooping on layer-3 switches in dual-stack networks.

Example commands:

```
// Enable on interface eth0
set protocols pim6 interface eth0

// Explicitly join multicast group ff18::1234 on interface eth1
set protocols pim6 interface eth1 mld join ff18::1234

// Explicitly join source-specific multicast group ff38::5678 with source address 2001:db8::1 on interface eth1
set protocols pim6 interface eth1 mld join ff38::5678 source 2001:db8::1
```

sever-sever commented on interface-definitions/protocols-mld.xml.in:
```diff
@@ -0,0 +1,95 @@
+                 <valueHelp>
+                    <format>2</format>
+                    <description>MLD version 2</description>
+                  </valueHelp>
```
Should there be  a default value? And which version is default?

vfreex replied:
Done adding the default value.

sever-sever commented on src/conf_mode/protocols_mld.py:
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2020-2023 VyOS maintainers and contributors
```
Just 2023

sever-sever commented on src/conf_mode/protocols_mld.py:
```diff
@@ -0,0 +1,125 @@
+        'ifaces'    : {}
+    }
+    if not (conf.exists('protocols mld') or conf.exists_effective('protocols mld')):
+        return None
```
The config exists and effective are old implementation of configuration. 
You should use get_config_dict ad in this example https://github.com/vyos/vyos-1x/blob/d3edda22573fb9c9d1c469f14f5a3eec9ca512a5/src/conf_mode/protocols_babel.py#L41

vfreex replied:
done

c-po commented:
Any reason to not make the daemon always run as we do with bgpd, ospfd, ospf6d?

vfreex commented:
> Any reason to not make the daemon always run as we do with bgpd, ospfd, ospf6d?

This PR mostly follows what `protocols igmp` does. I don't know if there is a reason actually. If we believe it is better to make the daemon always run I can make a change.

c-po commented on src/conf_mode/protocols_mld.py:
```diff
@@ -0,0 +1,99 @@
+airbag.enable()
+
+# Required to use the full path to pim6d, in another case daemon will not be started
+pim6d_cmd = f'/usr/lib/frr/pim6d -d -F traditional --daemon -A 127.0.0.1'
```
This is for IPv6, correct? Or should pim for IPv4 be used?

vfreex replied:
Yes, this is for IPv6. Only pim6d is needed for MLD.

c-po commented:
> Any reason to not make the daemon always run as we do with bgpd, ospfd, ospf6d?

Okay, this was me thinking it'sIPv4 PIM where we can only run pimd or igmp-proxy - but this is IPv6.

Can we make it `set protocols pim6` instead? So I see no reason to not include support for this in future versions.

c-po commented:
pim6d was enabled in the base image https://github.com/vyos/vyos-build/commit/1e10c349d4cbce933e6d6373c9e011e053ec3b31

Please use frr-reload like we do in other protocols (ospf, ospfv3 etc)

vfreex commented:
@c-po 

> Can we make it set protocols pim6 instead? So I see no reason to not include support for this in future versions.

Yes. This PR makes it in `set protocols mld` because we have `set protocols igmp` for IPv4 (I wanted to follow the same practice). Also this PR only adds MLD query and report support and doesn't include complete pim6 features in FRR.

> pim6d was enabled in the base image https://github.com/vyos/vyos-build/commit/1e10c349d4cbce933e6d6373c9e011e053ec3b31
> Please use frr-reload like we do in other protocols (ospf, ospfv3 etc)

Thanks. I will make a change to use frr-reload.

vfreex commented:
@c-po I've updated this PR and added smoketests for it. However I don't have an environment to run smoketests. Let me know if there are other any issues.

sever-sever commented on data/templates/frr/daemons.frr.tmpl:
```diff
@@ -38,6 +39,7 @@ isisd_options="  --daemon -A 127.0.0.1
 {%- if snmp is defined and snmp.isisd is defined %} -M snmp{% endif -%}
 "
 pimd_options="  --daemon -A 127.0.0.1"
+pim6d_options=""--daemon -A ::1"
```
Are there three double quotes?

vfreex replied:
Nice spot! Now fixed.

c-po commented on interface-definitions/protocols-pimv6.xml.in:
```diff
@@ -0,0 +1,105 @@
+<interfaceDefinition>
+  <node name="protocols">
+    <children>
+      <node name="pimv6" owner="${vyos_conf_scripts_dir}/protocols_pimv6.py">
```
Please use only `pim6` as FRR has `pim` and `pim6` no need for an extra `v`

vfreex replied:
Done

c-po commented on interface-definitions/protocols-pimv6.xml.in:
```diff
@@ -0,0 +1,105 @@
+                      <leafNode name="source">
+                        <properties>
+                          <help>Source address</help>
+                          <valueHelp>
```
Please add a completion helper:
```
    <completionHelp>
      <script>${vyos_completion_dir}/list_local_ips.sh --ipv6</script>
    </completionHelp>
```

vfreex replied:
Added

c-po commented on interface-definitions/protocols-pimv6.xml.in:
```diff
@@ -0,0 +1,105 @@
+                    </properties>
+                    <defaultValue>2</defaultValue>
+                  </leafNode>
+                  <leafNode name="query-interval">
```
Can we reduce this to `interval`? The help string is sufficient and we could try to make a common interval node.

vfreex replied:
renamed

c-po commented on interface-definitions/protocols-pimv6.xml.in:
```diff
@@ -0,0 +1,105 @@
+                      </constraint>
+                    </properties>
+                  </leafNode>
+                  <leafNode name="query-max-response-time">
```
Can we drop `query`?

Also deci seconds is an odd interval. As sub second granularity seems to be used - can we switch to milliseconds?

vfreex replied:
done. Also changed the unit to milliseconds and the value will be divided by 100 in the jinja template.

c-po commented on src/conf_mode/protocols_pimv6.py:
```diff
@@ -0,0 +1,102 @@
+    return pimv6
+
+
+def verify(pimv6: Optional[ConfigDict]):
```
Can we please stick with the current signature and not use Optional?

vfreex commented:
https://github.com/vyos/vyos-documentation/pull/1077 for documentation.
https://github.com/vyos/vyos-1x/pull/2233 fixes the smoketest.
